### PR TITLE
Ready to merge

### PR DESCRIPTION
**Pull request by arthbous**
Added newton backtrack in benchmark/PNP/pnp_EAFE.cpp

### CODE REVIEW

**Comment by arthbous** [comment]
I added the newton reading param to pnp_EAFE.cpp.
If that looks good to you, you can merge it and add the mesh refinement to pnp_eafe

**Comment by maximilianmetti** [comment]
Putting ``feature/`` in the name of the branch :+1:

**Comment by arthbous** [comment]
I learned from the best!

**Comment by maximilianmetti** [comment]
Can you put these updates into the unit tests too please?  Then, we can check whether they may break something in the future.

**Comment by arthbous** [comment]
I can, but we don't have it for coeff and all the other

**Comment by maximilianmetti** [comment]
If you could add it, that would be good.  Ultimately, the ``test_pnp`` and ``test_pnp_eafe`` will solve some simple PNP benchmark problems using our state-of-the-art implementation.  We need to keep updating them so that our tests are comprehensive.

File: benchmarks/PNP/pnp_EAFE.cpp [hunk 1]
```diff
@@ -31,7 +31,38 @@ double upper_cation_val = 1.0;  // 1 / m^3
 double lower_anion_val = 1.0;  // 1 / m^3
 double upper_anion_val = 0.1;  // 1 / m^3
 double lower_potential_val = -1.0;  // V
-double upper_potential_val = 1.0;  // V
+double upper_potential_val = 1.0;  //
+
+void update_solution_pnp(dolfin::Function& iterate0, dolfin::Function& iterate1, dolfin::Function& iterate2, \
+    dolfin::Function& update, double& relative_residual, pnp_and_source::LinearForm & L, const dolfin::DirichletBC bc, dolfin::EigenVector& b, double DampFactor, int maxDamp)
```
Comment (maximilianmetti) [inline benchmarks/PNP/pnp_EAFE.cpp]: why is ``relative_residual`` passed by reference here?

**Comment by arthbous** [comment]
because I update its value inside the function

File: benchmarks/PNP/pnp_EAFE.cpp [hunk 2]
```diff
@@ -31,7 +31,38 @@ double upper_cation_val = 1.0;  // 1 / m^3
...
+      i++;
+      L.CatCat = _iterate0;
+      L.AnAn   = _iterate1;
+      L.EsEs   = _iterate2;
+      assemble(b, L);
+      bc.apply(b);
+      new_residual = b.norm("l2");
+    }
+    *(iterate0.vector()) = *(_iterate0.vector()); *(iterate1.vector()) = *(_iterate1.vector()); *(iterate2.vector()) = *(_iterate2.vector());
+    relative_residual = new_residual;
```
Comment (maximilianmetti) [inline benchmarks/PNP/pnp_EAFE.cpp]: ``new_residual`` is not the relative_residual.  Isn't that what we should be computing here?  It looks like that's what should be returned

File: benchmarks/PNP/pnp_EAFE.cpp [hunk 3]
```diff
@@ -31,7 +31,38 @@ double upper_cation_val = 1.0;  // 1 / m^3
...
 double lower_potential_val = -1.0;  // V
-double upper_potential_val = 1.0;  // V
+double upper_potential_val = 1.0;  //
+
+void update_solution_pnp(dolfin::Function& iterate0, dolfin::Function& iterate1, dolfin::Function& iterate2, \
+    dolfin::Function& update, double& relative_residual, pnp_and_source::LinearForm & L, const dolfin::DirichletBC bc, dolfin::EigenVector& b, double DampFactor, int maxDamp)
+{
+    double new_residual=0.0;
+    int i=1;
+    dolfin::Function _iterate0(iterate0); dolfin::Function _iterate1(iterate1); dolfin::Function _iterate2(iterate2);
```
Comment (maximilianmetti) [inline benchmarks/PNP/pnp_EAFE.cpp]: I like the ``_`` to denote scope-restricted ``Function``s

**Comment by arthbous** [comment]
it's not the relative no, the division is commented.

you can add the line "relative_residual = relative_residual / initial_residual" in the main loop.

Don't add right now, let me finish unit test

**Comment by maximilianmetti** [comment]
It seems like this and the pnp-tests only check the absolute residual (not the relative!).  I'm going to create a PR to fix this and then I'll get back to this PR...

**Comment by arthbous** [comment]
Units test done and added you can change residual

**Comment by maximilianmetti** [comment]
We have to be a clever here. The relative residual grows at first and then converges... This is the typical behavior that I mentioned before (how Newton's method "wanders" before it converges).  I'm going to make it so that the relative residual cannot grow twice within a single iteration or 100 times overall...

**Comment by maximilianmetti** [comment]
@arthbous please run the unit tests and review the changes I make to benchmarks/pnp_eafe.cpp so that you find them agreeable.  Merge it if you like it; otherwise, let's get this to your liking.

**Pull request merged**